### PR TITLE
PLAT-9210 support collections when passing metadata to the Android SDK

### PR DESCRIPTION
**Pull request by rich-bugsnag**
### Add support for collections when passing metadata to the Android SDK

Previously all metadata was passed as strings to the android sdk, and collections were not passed at all.
Now all metadata is converted into the correct java type and passed through in a map so that all types are preserved and all collection types are supported.

### Improve metadata testing in E2E tests on all platforms

To test the new android functionality, metadata testing on all platforms has been improved.

## Changeset

- Rewrote the step `And custom metadata is included in the event` and the method `Scenario.AddTestingMetadata ` to test a wider variaty of data types.
- Rewrote the method `NativeInterface. DictionaryToJavaMap` to support multiple types of data and added support for null metadata.
- Changed the method `NativeInterface. AddMetadata` to pass the metadata as a java map instead of passing each property as a string one at a time. 

### This requires a change in mazerunner to support validating float and double types
- Added tighter metadata testing

### CODE REVIEW

File: features/steps/unity_steps.rb [hunk 1]
```diff
@@ -184,18 +184,24 @@ def check_error_reporting_api(notifier_name)
...
-    And the event "metaData.init" is null
-    And the event "metaData.custom.letter" equals "QX"
-    And the event "metaData.custom.better" equals 400
-    And the event "metaData.test.test1" equals "test1"
-    And the event "metaData.test.test2" is null
-    And the error payload field "events.0.metaData.custom.int-array" is a non-empty array
-    And the error payload field "events.0.metaData.custom.string-array" is a non-empty array
-    And the error payload field "events.0.metaData.custom.dict" is not null
+    Then the event "metaData.custom.int" equals 123
+    And the event "metaData.custom.float" equals "123.123"
```
Comment (twometresteve) [inline features/steps/unity_steps.rb:193]: You'll need to lose the quote around the number.

File: features/fixtures/maze_runner/Assets/Scripts/Scenario.cs [hunk 2]
```diff
@@ -87,24 +87,37 @@ private static string FindDotnetApiCompatibility()
...
+
+        Bugsnag.ClearMetadata("clearMe");
+
+        Bugsnag.AddMetadata("test", "test1", "test1");
+        Bugsnag.AddMetadata("test", "test1", "test2");
+        Bugsnag.AddMetadata("test", "nullMe", "notNull");
+        Bugsnag.AddMetadata("test", "nullMe", null);
+
         Bugsnag.AddMetadata("app", new Dictionary<string, object>(){
-            {"buildno", "0.1" },
```
Comment (twometresteve) [inline features/fixtures/maze_runner/Assets/Scripts/Scenario.cs:103]: The removal of this appears to be causing the following (embedded) step to fail in unty_steps.rb:
```
    And the event "metaData.app.buildno" is not null
```

Comment (rich-bugsnag) [reply features/fixtures/maze_runner/Assets/Scripts/Scenario.cs:103]: thanks 👍

**Review by twometresteve — COMMENTED** [review]
Test aspects LGTM, just asked one question and will let others approve the dev side.

File: features/fixtures/maze_runner/Assets/Scripts/Scenario.cs [hunk 3]
```diff
@@ -87,24 +87,37 @@ private static string FindDotnetApiCompatibility()
...
-            {"letter", "QX" },
-            {"better", 400 },
-            {"string-array", new string []{"1","2","3"} },
-            {"int-array", new int []{1,2,3} },
-            {"dict", new Dictionary<string,object>(){ {"test" , 123 } } }
+            {"int", 123 },
+            {"float", 123.123f },
+          //  {"long", 12345678901234567890 }, pending PLAT-9426
+            {"double", 123.456 },
+            {"stringArray", new []{"1",null,"3"} },
```
[the remark this answers is not in the document]
Comment (rich-bugsnag) [reply features/fixtures/maze_runner/Assets/Scripts/Scenario.cs:96]: done 👍

**Review by lemnik — APPROVED** [review]
LGTM

**Pull request merged**